components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/saml/SAML2BearerGrantHandler.java

```diff
@@ -128,476 +131,628 @@ public void init() throws IdentityOAuth2Exception {
+            throws IdentityProviderManagementException {
+        return IdentityProviderManager.getInstance().getIdPByAuthenticatorPropertyValue(IDP_ENTITY_ID,
+                        assertion.getIssuer().getValue(), tenantDomain, authenticatorProperty, false);
+    }
 
-        /*
-          The authorization server MUST verify that the Assertion is valid in all other respects per
-          [OASIS.saml-core-2.0-os], such as (but not limited to) evaluating all content within the Conditions
-          element including the NotOnOrAfter and NotBefore attributes, rejecting unknown condition types, etc.
+    private void checkNullIdentityProvider(Assertion assertion, String tenantDomain, IdentityProvider identityProvider)
+            throws IdentityOAuth2Exception {
+        if (identityProvider == null) {
+            if(log.isDebugEnabled()) {
+                log.debug("SAML Token Issuer : " + assertion.getIssuer().getValue() +
+                        " not registered as a local Identity Provider in tenant : " + tenantDomain);
+            }
+            throw new IdentityOAuth2Exception("Identity provider is null");
+        }
+    }
 
-          [OASIS.saml-core-2.0-os] - http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf
-         */
+    private boolean validateIssuer(OAuthTokenReqMessageContext tokReqMsgCtx, Assertion assertion)
+            throws IdentityOAuth2Exception {
+        if (issuerNotFoundInAssertion(assertion)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Issuer is empty in the SAML assertion. Token request for user : " +
+                        tokReqMsgCtx.getAuthorizedUser());
+            }
+            throw new IdentityOAuth2Exception("Issuer is empty in the SAML assertion");
+        }
+        return true;
+    }
 
+    private boolean issuerNotFoundInAssertion(Assertion assertion) {
+        return assertion.getIssuer() == null || StringUtils.isEmpty(assertion.getIssuer().getValue());
+    }
 
-        // TODO: Throw the SAML request through the general SAML2 validation routines
+    /**
+     * The Assertion MUST contain a <Subject> element.  The subject MAY identify the resource owner for whom
+     * the access token is being requested.  For client authentication, the Subject MUST be the "client_id"
+     * of the OAuth client.  When using an Assertion as an authorization grant, the Subject SHOULD identify
+     * an authorized accessor for whom the access token is being requested (typically the resource owner, or
+     * an authorized delegate).  Additional information identifying the subject/principal of the transaction
+     * MAY be included in an <AttributeStatement>.
+     * @param tokReqMsgCtx
+     * @param assertion
+     * @throws IdentityOAuth2Exception
+     */
+    private boolean validateSubject(OAuthTokenReqMessageContext tokReqMsgCtx, Assertion assertion)
+            throws IdentityOAuth2Exception {
+        if (assertion.getSubject() != null) {
```

Comment:
What happens if the subject is empty. Shall we use isBlank?

Reply:
assertion.getSubject() returns a Subject object. This if condition is to check whether that Subject object is null.